src/utils/filter-utils.js

```diff
@@ -134,33 +143,66 @@ export function getDefaultFilter(dataId) {
  * @return {boolean} true if a filter is valid, false otherwise
  */
 export function shouldApplyFilter(filter, datasetId) {
-  const valid = Boolean(filter.dataId)
-    // datasetId is contained in dataId
-    && (
-      // filter.dataId is an array or or matches exactly the datasetId
-      (Array.isArray(filter.dataId) && filter.dataId.includes(datasetId))
-      || filter.dataId === datasetId
-    )
-    // value must be defined
-    && filter.value !== null;
-
-  return valid;
+  const dataIds = toArray(filter.dataId);
+  return dataIds.includes(datasetId) && filter.value !== null;
 }
 
 /**
- * Validate saved filter config with new data,
- * calculate domain and fieldIdx based new fields and data
- *
- * @param {Object} dataset
- * @param {Object} filter - filter to be validate
- * @return {Object | null} - validated filter
+ * Validates and modifies polygon filter structure
+ * @param dataset
+ * @param filter
+ * @param layers
+ * @return {object}
  */
-export function validateFilterWithData(dataset, filter) {
+export function validatePolygonFilter(dataset, filter, layers) {
+  const failed = {dataset, filter: null};
+  const {value, layerId, type, dataId} = filter;
+
+  if (!layerId || !isValidFilterValue(type, value)) {
+    return failed;
+  }
+
+  const isValidDataset = dataId.includes(dataset.id);
 
+  if (!isValidDataset) {
+    return failed;
+  }
+
+  const layer = layers.find(l => layerId.includes(l.id));
+
+  if (!layer) {
+    return failed;
+  }
+
+  return {
+    filter: {
+      ...filter,
+      freeze: true,
+      fieldIdx: [0]
```

Comment:
why `fieldIdx` needs to be [0]? if there is no field selected, can't it be `[]`?